pallets/block-rewards/src/lib.rs

```diff
@@ -0,0 +1,449 @@
+
+		#[pallet::constant]
+		type CollatorCurrencyId: Get<Self::CurrencyId> + TypeInfo;
+
+		#[pallet::constant]
+		type CollatorGroupId: Get<Self::GroupId> + TypeInfo;
+
+		#[pallet::constant]
+		type DefaultCollatorStake: Get<Self::Balance> + TypeInfo;
+
+		#[pallet::constant]
+		type MaxCollators: Get<u32> + TypeInfo + sp_std::fmt::Debug + Clone + PartialEq;
+
+		/// Target of receiving non-collator-rewards.
+		/// NOTE: If set to none, collators are the only group receiving rewards.
+		type Beneficiary: OnUnbalanced<NegativeImbalanceOf<Self>>;
+
+		/// The identifier type for an authority.
+		type AuthorityId: Member
+			+ Parameter
+			+ sp_runtime::RuntimeAppPublic
+			+ MaybeSerializeDeserialize
+			+ MaxEncodedLen;
+
+		/// Information of runtime weights
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	/// Data associated to the current epoch.
+	#[pallet::storage]
+	pub(super) type ActiveEpochData<T: Config> = StorageValue<_, EpochData<T>, ValueQuery>;
+
+	/// Pending update data used when the current epoch finalizes.
+	/// Once it's used for the update, it's reset.
+	#[pallet::storage]
+	pub(super) type NextEpochChanges<T: Config> = StorageValue<_, EpochChanges<T>, ValueQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		NewEpoch {
+			collator_reward: T::Balance,
+			beneficiary_reward: T::Balance,
+			last_changes: EpochChanges<T>,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Limit of max calls with same id to [`Pallet::set_group_weight()`] or
+		/// [`Pallet::set_currency_group()`] reached.
+		MaxChangesPerEpochReached,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Claims the reward the associated to a currency.
+		/// The reward will be transferred to the target account.
+		#[pallet::weight(T::WeightInfo::claim_reward())]
+		#[transactional]
+		pub fn claim_reward(
+			origin: OriginFor<T>,
+			currency_id: T::CurrencyId,
+			account_id: T::AccountId,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			T::Rewards::claim_reward((T::Domain::get(), currency_id), &account_id).map(|_| ())
+		}
+
+		/// Admin method to set the reward amount for a collator used for the next epochs.
+		/// Current epoch is not affected by this call.
+		#[pallet::weight(T::WeightInfo::set_distributed_reward())]
+		pub fn set_collator_reward(
+			origin: OriginFor<T>,
+			collator_reward_per_session: T::Balance,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			NextEpochChanges::<T>::mutate(|changes| {
+				changes.collator_reward = Some(collator_reward_per_session);
+			});
+
+			Ok(())
+		}
+
+		/// Admin method to set the inflation amount for the Beneficiary used for the next epochs.
+		/// Current epoch is not affected by this call.
+		#[pallet::weight(T::WeightInfo::set_distributed_reward())]
+		pub fn set_beneficiary_reward(
```

Comment:
We could definitely do it your way as well! I kind of went this path before [1ee21af](https://github.com/centrifuge/centrifuge-chain/pull/1198/commits/1ee21af95af8a31861dfd2ca22c08f9812777c67), when it became aware that rewarding a group with amount `x` consumes the entire `x`. 

> Everytime we include a new collator, we will need to update this value, which possibly for a period of time we would be giving more than what we should.

The current design does not need to be reconfigured when new collators join as the configured reward is per collator, at least under the assumption that `n * collator_reward + beneficiary_amount` does not need to stay static.
However, I definitely see the upside in your approach. Just on a note: We would still need to store two reward amounts, otherwise we would have to hardcode the share for a collator.

Reply:
yes, the `collator_reward` is per collator and that doesn't have to change when changing the collator size, but that new value now affects the `beneficiary_amount` that goes into the treasury right? That value will need to be adapted accordingly when a new size of collators is defined. 
Am I understanding correctly that the `beneficiary_reward` is the `3% inflation rate - the collator reward`? If so those two values are interconnected.
Yes, in my proposal we still have to store 2 values.

Reply:
Yes that is correct. Let's summarize: I will change `beneficiary_reward` to denote the total reward per epoch instead. From that, we subtract the rewards for all collators and the rest
* goes to the Treasury for Centrifuge chain
* stays in the Treasury for Altair

Reply:
I think so, that means that the `beneficiary_reward`, if set, has to be at least the same as the collator_reward * number of collators at that point in time. 
And for Altair, we basically do not set it, and the logic should be able to only mint/move the collator_reward without interacting with the Zero value of the `beneficiary_reward`.

Reply:
Fixed in [8b8c40f](https://github.com/centrifuge/centrifuge-chain/pull/1198/commits/8b8c40f805b8c8b6b1c0eed9b0ffa00df17e6ff3). Not sure if I should rename `Beneficiary` to `Treasury`.